vowpalwabbit/learner.h

```diff
@@ -191,139 +220,280 @@ typedef void (*tend_example)(vw& all, void* d, example& ec);
   { end_pass_fd.func(end_pass_fd.data);
     if (end_pass_fd.base) end_pass_fd.base->end_pass();
   }//autorecursive
-  void set_end_pass(void (*f)(T&))
-  {end_pass_fd = tuple_dbf(learn_fd.data, learn_fd.base, (tfunc)f);}
+  void set_end_pass(void (*f)(T&)) {end_pass_fd = tuple_dbf(learn_fd.data, learn_fd.base, (func_data::fn)f);}
 
   //called after parsing of examples is complete.  Autorecursive.
   void end_examples()
   { end_examples_fd.func(end_examples_fd.data);
     if (end_examples_fd.base) end_examples_fd.base->end_examples();
   }
-  void set_end_examples(void (*f)(T&))
-  {end_examples_fd = tuple_dbf(learn_fd.data,learn_fd.base, (tfunc)f);}
+  void set_end_examples(void (*f)(T&)) {end_examples_fd = tuple_dbf(learn_fd.data,learn_fd.base, (func_data::fn)f);}
 
   //Called at the beginning by the driver.  Explicitly not recursive.
   void init_driver() { init_fd.func(init_fd.data);}
   void set_init_driver(void (*f)(T&))
-  { init_fd = tuple_dbf(learn_fd.data,learn_fd.base, (tfunc)f); }
+  { init_fd = tuple_dbf(learn_fd.data,learn_fd.base, (func_data::fn)f); }
 
   //called after learn example for each example.  Explicitly not recursive.
-  inline void finish_example(vw& all, example& ec)
-  { finish_example_fd.finish_example_f(all, finish_example_fd.data, ec);}
-  void set_finish_example(void (*f)(vw& all, T&, example&))
+  inline void finish_example(vw& all, E& ec)
+  { finish_example_fd.finish_example_f(all, finish_example_fd.data, (void*)&ec);}
+  //called after learn example for each example.  Explicitly not recursive.
+  void set_finish_example(void (*f)(vw& all, T&, E&))
   { finish_example_fd.data = learn_fd.data;
-    finish_example_fd.finish_example_f = (tend_example)f;
+    finish_example_fd.finish_example_f = (end_fptr_type)(f);
+  }
+  
+  //called to check if an example is a test example.  Used in creating multiline example
+  inline bool is_test_example(example& ec)
```

Comment:
Again, this belongs to the label type not the learner.

Reply:
Defered to John's upcoming change.